govcd/vm.go

```diff
@@ -850,3 +851,186 @@ func (vm *VM) SetGuestCustomizationSection(guestCustomizationSection *types.Gues
 
 	return vm.GetGuestCustomizationSection()
 }
+
+// AddInternalDisk creates disk type *types.DiskSettings to the VM.
+// Returns new disk ID and error.
+// Runs synchronously, VM is ready for another operation after this function returns.
+func (vm *VM) AddInternalDisk(diskData *types.DiskSettings) (string, error) {
+	err := vm.Refresh()
+	if err != nil {
+		return "", fmt.Errorf("error refreshing VM: %s", err)
+	}
+
+	err = vm.validateInternalDiskInput(diskData)
+	if err != nil {
+		return "", err
+	}
+
+	var diskSettings []*types.DiskSettings
+	if vm.VM.VmSpecSection != nil && vm.VM.VmSpecSection.DiskSection != nil && vm.VM.VmSpecSection.DiskSection.DiskSettings != nil {
+		diskSettings = vm.VM.VmSpecSection.DiskSection.DiskSettings
+	}
+
+	diskSettings = append(diskSettings, diskData)
+	vmSpecSection := vm.VM.VmSpecSection
+	vmSpecSection.DiskSection.DiskSettings = diskSettings
+
+	vmSpecSection, err = vm.UpdateInternalDisks(vmSpecSection)
+	if err != nil {
+		return "", err
+	}
+
+	for _, diskSetting := range vmSpecSection.DiskSection.DiskSettings {
+		if diskSetting.AdapterType == diskData.AdapterType &&
+			diskSetting.BusNumber == diskData.BusNumber &&
+			diskSetting.UnitNumber == diskData.UnitNumber {
+			return diskSetting.DiskId, nil
+		}
+	}
+
+	return "", fmt.Errorf("created disk wasn't in list of returned VM internal disks")
+}
+
+func (vm *VM) validateInternalDiskInput(diskData *types.DiskSettings) error {
+	if diskData.AdapterType == "" {
+		return fmt.Errorf("disk settings missing required field: adapter type")
+	}
+
+	if diskData.BusNumber < 0 {
+		return fmt.Errorf("disk settings bus number has to be 0 or higher")
+	}
+
+	if diskData.UnitNumber < 0 {
+		return fmt.Errorf("disk settings unit number has to be 0 or higher")
+	}
+
+	if diskData.SizeMb < int64(0) {
+		return fmt.Errorf("disk settings size MB has to be 0 or higher")
+	}
+
+	if diskData.Iops != nil && *diskData.Iops < int64(0) {
+		return fmt.Errorf("disk settings iops has to be 0 or higher")
+	}
+
+	if diskData.ThinProvisioned == nil {
+		return fmt.Errorf("disk settings missing required field: thin provisioned")
+	}
+
+	if diskData.StorageProfile == nil {
+		return fmt.Errorf("disk settings missing required field: storage profile")
+	}
+
+	return nil
+}
+
+// GetInternalDiskById returns a valid *types.DiskSettings if it exists.
+// If it doesn't, returns nil and ErrorEntityNotFound or other err.
+func (vm *VM) GetInternalDiskById(diskId string, refresh bool) (*types.DiskSettings, error) {
+	if refresh {
+		err := vm.Refresh()
+		if err != nil {
+			return nil, fmt.Errorf("error refreshing VM: %s", err)
+		}
+	}
+
+	if diskId == "" {
```

Comment:
We should check for valid ID before refreshing.

Reply:
Moved.